src/ui/map.js

```diff
@@ -2171,22 +2171,58 @@ class Map extends Camera {
     }
 
     /**
-     * Queries the currently loaded data for elevation at a geographical location. This accounts for the value of `exaggeration` set on `terrain`.
+     * Queries the currently loaded data for elevation at a geographical location. The elevation is returned in `meters` and accounts for the value of `exaggeration` set on `terrain`.
      * Returns `null` if `terrain` is disabled or if terrain data for the location hasn't been loaded yet.
      *
      * In order to guarantee that the terrain data is loaded ensure that the geographical location is visible and wait for the `idle` event to occur.
-     * @param {LngLatLike} lnglat The geographical location to project.
+     * @param {LngLatLike} lnglat The geographical location at which to query.
      * @returns {number | null} The elevation in meters, accounting for `terrain.exaggeration`.
      * @example
      * var coordinate = [-122.420679, 37.772537];
-     * var elevation = map.queryTerrainElevationAt(coordinate);
+     * var elevation = map.queryTerrainElevationAtLocation(coordinate);
      */
-    queryTerrainElevationAt(lnglat: LngLatLike): number | null {
+    queryTerrainElevationAtLocation(lnglat: LngLatLike): number | null {
         if (!this.transform.elevation) return null;
 
         return this.transform.elevation.getAtPoint(MercatorCoordinate.fromLngLat(lnglat));
     }
 
+    /**
+     * Queries the visible data for elevation at a geographical point that corresponds to the specified pixel coordinates. The elevation is returned in `meters` and accounts for the value of `exaggeration` set on `terrain`.
+     * Returns `null` if `terrain` is disabled, if terrain data for the location hasn't been loaded yet, or when the specified point is above the horizon.
+     *
+     * In order to guarantee that terrain data is loaded wait for the `idle` event to occur.
+     * @param {PointLike} point The pixel coordinates at which to query.
+     * @returns {number | null} The elevation in meters, accounting for `terrain.exaggeration`.
+     * @example
+     * map.on('mousemove', function(e) {
+     *   // Get the elevation underneath the mouse pointer
+     *   var elevation = map.queryTerrainElevation(e.point);
+     * });
+     */
+    queryTerrainElevation(point: PointLike): number | null {
```

Comment:
Why not to returning also `LngLat` of the raycast point, I mean in addition to elevation?

Reply:
We already have `unproject` for that. Initially I was actually thinking of adding `altitude` or `z` to `LngLat` itself and not add new methods, but that made things a bit tricky.

For all out other functions that take a `LngLat` (`project`, `map.getCenter()/setCenter(), `easeTo()` etc) we have already assumed that ` LngLat` is at the ground surface and we automatically interpret its altitude by looking at the height underneath the terrain for that supplied `LngLat`. If we add z encoded into the LngLat itself we then need another mechanism to disambiguate between `LngLat.z` and the terrain z at that location.

So I felt it was simpler to keep the assumption that a `LngLat` is always on the ground/terrain surface like we have currently and instead add a new method to extract the z.

Reply:
This got addressed later - original implementation here was using transform.elevation.pointCoordinate and not clamping points above ground (clamping to ground done in unproject using transform.pointCoordinate3D).

While `unproject` is a best effort to seamlessly  integrate skybox and terrain to existing application code, there is no raycast API exposed that wouldn't clamp.